ksqldb-functional-tests/src/test/resources/query-validation-tests/greatest.json

```diff
@@ -0,0 +1,139 @@
+{
+  "comments": [
+    "Tests covering the greatest function."
+  ],
+  "tests": [
+    {
+      "name": "test greatest with numeric types",
+      "format": ["JSON"],
+      "statements": [
+        "CREATE STREAM INPUT (ID BIGINT KEY, N1 INT, N2 INT, N3 INT, N4 INT) WITH (kafka_topic='input_topic',value_format='{FORMAT}');",
+        "CREATE STREAM OUTPUT AS SELECT ID, GREATEST(N1, N2, N3, N4, null, null, 5) AS HIGHEST FROM INPUT;"
+      ],
+      "inputs": [
+        {"topic": "input_topic", "key": 1, "value": {"N1": 6, "N2": -1, "N3":  0, "N4": null}},
+        {"topic": "input_topic", "key": 2, "value": {"N1": 2147483647, "N2": -2147483648, "N3":  0, "N4": null}},
+        {"topic": "input_topic", "key": 3, "value": {"N1": null, "N2": null, "N3":  null, "N4": null}}
+      ],
+      "outputs": [
+        {"topic": "OUTPUT", "key": 1, "value": {"HIGHEST": 6}},
+        {"topic": "OUTPUT", "key": 2, "value": {"HIGHEST": 2147483647}},
+        {"topic": "OUTPUT", "key": 3, "value": {"HIGHEST": 5}}
+      ]
+    },
+    {
+      "name": "test implicit casting to long",
+      "format": ["JSON"],
+      "statements": [
+        "CREATE STREAM INPUT (ID BIGINT KEY, N1 INT, N2 BIGINT) WITH (kafka_topic='input_topic',value_format='{FORMAT}');",
+        "CREATE STREAM OUTPUT AS SELECT ID, GREATEST(N1, N2) AS HIGHEST FROM INPUT;"
+      ],
+      "expectedException": {
+        "type": "io.confluent.ksql.util.KsqlStatementException",
+        "message": "Function 'greatest' does not accept parameters (INTEGER, BIGINT)."
+      }
+    },
+    {
+      "name": "test implicit casting to decimal",
+      "format": ["JSON"],
+      "statements": [
+        "CREATE STREAM INPUT (ID BIGINT KEY, N1 INT, N2 BIGINT, N3 DECIMAL(50,30)) WITH (kafka_topic='input_topic',value_format='{FORMAT}');",
+
+        "CREATE STREAM OUTPUT AS SELECT ID, GREATEST(N1, N2, N3, null) AS HIGHEST FROM INPUT;"
+
+      ],
+      "expectedException": {
+        "type": "io.confluent.ksql.util.KsqlStatementException",
+        "message": "Function 'greatest' does not accept parameters (INTEGER, BIGINT, DECIMAL(50, 30), null)."
+      }
+    },
+    {
+      "name": "test implicit casting to double",
```

Comment:
Why are implicit casts allowed to double but not to long or decimal? Is it because if there's at least one double then the only variant of the function that is applicable is the one that accepts doubles, and therefore the implicit cast of the other columns is allowed but this is not the case for the other variants? If so it'd be good to check with @colinhicks that this implicit cast behavior is desirable. It feels unintuitive to me as a user that some implicit casts are allowed (namely, anything involving at least one double) while others are not (i.e., all other combinations).

Also, we should add tests to exercise the other variants of the method since the only ones being tested currently are the int variant and the double variant.

Reply:
> Thanks @Sullivan-Patrick -- great job tracking down the UdfIndex fix! What do you think about separating the UdfIndex fix into its own PR? The UdfIndex fix is required in order to merge LEAST and GREATEST but conceptually it's separate.

Because I'm embarrassed how long it'll take me to do that on git :D but in all seriousness, sure. When it was taking 1 line from the compare() method it was probably more in the line of "just add it on top of what I have" but I did continue to add on it and now I agree, it warrants it's own PR.

> 
> Some general reasons it's good practice to separate PRs into unit-sized pieces:
> 
> * easier to rollback individual changes without impacting others (doesn't strictly apply in this case since one piece depends on another)
> * cleaner and more easily searchable commit histories
> * easier for reviewers (and the author!) to think about the design and test coverage for each change -- for example, separating out the UdfIndex changes to be in their own PR makes it easy to spot that we're currently missing unit test coverage for these changes. We should add some :)
> 
> Another additional concrete reasons to separate out the UdfIndex change from the PR that adds LEAEST and GREATEST: the UdfIndex change is a breaking change. It's a bit weird for someone to see in the changelog that a PR titled "feat: add least and greatest UDFs" contains a breaking change; from the title, it's unclear what the breaking change could be.
> 
> That said, we should also be practical about the number of PRs we're opening. It's totally fine to wrap small fixes in with other (ideally related) PRs, but in this case I think the change is large enough it merits its own PR.

Reply:
> Why are implicit casts allowed to double but not to long or decimal? Is it because if there's at least one double then the only variant of the function that is applicable is the one that accepts doubles, and therefore the implicit cast of the other columns is allowed but this is not the case for the other variants?

Yep, exactly.

>If so it'd be good to check with @colinhicks that this implicit cast behavior is desirable. It feels unintuitive to me as a user that some implicit casts are allowed (namely, anything involving at least one double) while others are not (i.e., all other combinations).

Ok. Should I slack him. or will your ping here suffice to summon him?


> Also, we should add tests to exercise the other variants of the method since the only ones being tested currently are the int variant and the double variant.

I'll add that to my first test, I forgot to edit after changing the ambiguous casting rules.

Reply:
I agree the implicit cast behavior is not intuitive, but can we offer better option with a reasonable level of effort?

Reply:
The only way I can think of to solve this would be to prevent implicit casting to double as well, which would require some uncomfortably case-specific logic inside `UdfIndex`. At that point I'd almost change the logic to trying to conform to the narrowest common numeric type (which I'm honestly happy to revisit after KLIP-50 is done, but might not be the best thing to take on now). I think while this behavior seems strange in isolation (given that I'm changing the behavior of `UdfIndex` in the same stroke as adding these UDFs), this behavior will make more, albeit still not perfect, sense in future cases.

I lean towards keeping it as for now, but I obviously defer to those with far more experience than me to make this decision (or put forward an alternative I haven't even considered)

Reply:
Yeah, let's not jump through extra hoops here then. @colinhicks what do you think about adding a note in the docs for these new functions to suggest users always cast the columns to be of the same type, and sweep the fact that implicitly casting works with doubles under the rug? I think it's easier to explain "always cast so your columns are the same type" than "for most type mismatches you'll have to do the cast yourself but there are some situations in which you don't."

Reply:
> what do you think about adding a note in the docs for these new functions to suggest users always cast the columns to be of the same type

SGTM